app/code/community/Bolt/Boltpay/Helper/Coupon.php

```diff
@@ -0,0 +1,662 @@
+<?php
+/**
+ * Bolt magento plugin
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * http://opensource.org/licenses/osl-3.0.php
+ *
+ * @category   Bolt
+ * @package    Bolt_Boltpay
+ * @copyright  Copyright (c) 2018 Bolt Financial, Inc (https://www.bolt.com)
+ * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+
+/**
+ * Class Bolt_Boltpay_Helper_Coupon
+ *
+ * Base Magento Bolt Coupon Helper class
+ *
+ */
+class Bolt_Boltpay_Helper_Coupon extends Mage_Core_Helper_Abstract
+{
+    const ERR_INSUFFICIENT_INFORMATION = 6200;
+    const ERR_CODE_INVALID = 6201;
+    const ERR_CODE_EXPIRED = 6202;
+    const ERR_CODE_NOT_AVAILABLE = 6203;
+    const ERR_CODE_LIMIT_REACHED = 6204;
+    const ERR_MINIMUM_CART_AMOUNT_REQUIRED = 6205;
+    const ERR_UNIQUE_EMAIL_REQUIRED = 6206;
+    const ERR_ITEMS_NOT_ELIGIBLE = 6207;
+    const ERR_SERVICE = 6001;
+
+    /**
+     * Response information sent to bolt
+     */
+    protected $responseStatus; // Bolt error status code
+    protected $responseError; // Error description
+    protected $responseCart; // Response quote
+    protected $discountInfo; // Details about how much discount was applied as a result of coupon
+    protected $httpCode; // HTTP response code
+
+    protected $immutableQuote = null;
+    protected $parentQuote = null; // Session quote
+    protected $coupon = null;
+    protected $rule = null;
+
+    protected $requestObject = null;
+
+    /**
+     * Applies coupon from hook input to customer quote if the coupon is valid.
+     *
+     * @return bool Whether or not coupon was applied
+     */
+    public function applyCoupon()
+    {
+        try {
+            $this->setupVariables();
+            $this->validateCoupon();
+            $this->applyCouponToQuotes();
+            $this->validateAfterApplyingCoupon();
+
+            $this->setSuccessResponse();
+        } catch (\Exception $e) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Sets up any variables needed to process coupon.
+     *
+     * @throws Exception     thrown if input fails to be read.
+     */
+    protected function setupVariables() {
+        try {
+            $this->requestObject = json_decode(file_get_contents('php://input'));
+        } catch (\Exception $e) {
+            $this->setErrorResponseAndThrowException(
+                self::ERR_SERVICE,
+                'Unknown error getting API request.',
+                422
+            );
+
+            throw $e;
```

Comment:
since you call `setErrorResponseAndThrowException` there is no need to re-throw here?